EIPS/eip-2390.md

```diff
@@ -0,0 +1,314 @@
+See https://github.com/james-choncholas/resolvers/blob/master/test/TestPublicResolver.js
+
+## Implementation
+<!--The implementations must be completed before any EIP is given status "Final", but it need not be completed before the EIP is accepted. While there is merit to the approach of reaching consensus on the specification and rationale before writing code, the principle of "rough consensus and running code" is still useful when it comes to resolving many discussions of API details.-->
+
+This address resolver, written in Solidity, implements the specifications outlined above.
+The same idea presented here can be applied to other resolver interfaces as specified in EIP137.
+Note that geohashes are passed and stored using 64 bit unsigned integers.
+Using integers instead of strings for geohashes is more performant, especially in the `geomap` mapping.
+For comparison purposes, see https://github.com/james-choncholas/geoens/tree/master/contracts/StringOwnedGeoENSResolver.sol for the inefficient string implementation.
+
+
+```solidity
+pragma solidity ^0.5.0;
+
+import "../ResolverBase.sol";
+
+contract GeoENSResolver is ResolverBase {
+    bytes4 constant ERC2390 = 0xa263115e;
+    uint constant MAX_ADDR_RETURNS = 64;
+    uint constant TREE_VISITATION_QUEUESZ = 64;
+    uint8 constant ASCII_0 = 48;
+    uint8 constant ASCII_9 = 57;
+    uint8 constant ASCII_a = 97;
+    uint8 constant ASCII_b = 98;
+    uint8 constant ASCII_i = 105;
+    uint8 constant ASCII_l = 108;
+    uint8 constant ASCII_o = 111;
+    uint8 constant ASCII_z = 122;
+
+    struct Node {
+        address data; // 0 if not leaf
+        uint256 parent;
+        uint256[] children; // always length 32
+    }
+
+    // A geohash is 8, base-32 characters.
+    // A geomap is stored as tree of fan-out 32 (because
+    // geohash is base 32) and height 8 (because geohash
+    // length is 8 characters)
+    mapping(bytes32=>Node[]) private geomap;
+
+    event GeoENSRecordChanged(bytes32 indexed node, string geohash, address addr);
+
+    // only 5 bits of ret value are used
+    function chartobase32(byte c) pure internal returns (uint8 b) {
+        uint8 ascii = uint8(c);
+        require( (ascii >= ASCII_0 && ascii <= ASCII_9) ||
+                (ascii > ASCII_a && ascii <= ASCII_z));
+        require(ascii != ASCII_a);
+        require(ascii != ASCII_i);
+        require(ascii != ASCII_l);
+        require(ascii != ASCII_o);
+
+        if (ascii <= (ASCII_0 + 9)) {
+            b = ascii - ASCII_0;
+
+        } else {
+            // base32 b = 10
+            // ascii 'b' = 0x60
+            // note base32 skips the letter 'a'
+            b = ascii - ASCII_b + 10;
+
+            // base32 also skips the following letters
+            if (ascii > ASCII_i)
+                b --;
+            if (ascii > ASCII_l)
+                b --;
+            if (ascii > ASCII_o)
+                b --;
+        }
+        require(b < 32); // base 32 cant be larger than 32
```

Comment:
```suggestion
        require(b < 32); // base 32 can't be larger than 32
```

Reply:
The bot won't let this be merged until the spelling error is fixed.  Unfortunately, we don't have an easy way to tell the bot to ignore a particular spelling error (such as one embedded in code).

Reply:
Whoops - my mistake! I fixed at one point but the mistake crept back in when I changed the API from strings to bytes